django-rgd-3d/rgd_3d/tasks/etl.py

```diff
@@ -78,3 +86,86 @@ def read_mesh_3d_file(mesh_3d):
+    # See https://github.com/CesiumGS/3d-tiles/tree/main/specification#coordinate-reference-system-crs
+    srid = 4978
+    if 'region' in volume:
+        xmin, ymin, xmax, ymax, _, _ = volume['region']
+        # The region bounding volume specifies bounds using a geographic coordinate system (latitude, longitude, height), specifically EPSG 4979.
+        # See https://github.com/CesiumGS/3d-tiles/tree/main/specification#coordinate-reference-system-crs
+        srid = 4979
+    elif 'box' in volume:
+        b = volume['box']
+        center, x, y = b[0:3], b[3:6], b[6:9]
+        xmin = min(center[0], x[0], y[0])
+        xmax = max(center[0], x[0], y[0])
+        ymin = min(center[1], x[1], y[1])
+        ymax = max(center[1], x[1], y[1])
+        # TODO: for the oriented bounding box, use it as the `footprint` field
+    elif 'sphere' in volume:
+        raise NotImplementedError
+    else:
+        raise ValueError(f'Bounding volume of unknown type: {volume}')
+
+    coords = np.array(
+        [
+            (xmin, ymax),
+            (xmax, ymax),
+            (xmax, ymin),
+            (xmin, ymin),
+            (xmin, ymax),  # Close the loop
+        ]
+    )
+
+    if 'transform' in tileset_json['root']:
+        # Apply transform if required
+        transform = tileset_json['root']['transform']
+        # The transform is stored in column-major order in the JSON, so reorder
+        # to produce the  full 3D Affine matrix in the form of:
+        #      [x']    / a  b  c xoff \ [x]
+        #      [y'] =  | d  e  f yoff | [y]
+        #      [z']    | g  h  i zoff | [z]
+        #      [1 ]    \ 0  0  0   1  / [1]
+        t = np.array(transform).reshape((4, 4), order='F')
+        # But we want the 2D subcomponent which would be:
+        #      [x']    / a  b xoff \ [x]
+        #      [y'] =  | d  e yoff | [y]
+        #      [1 ]    \ 0  0   1  / [1]
+        # This would effectively be: np.hstack((t[0:2,0:2], t[0:2,-1,None]))
+        # but, shapely expects the following form as a 1D array:
+        #           [a, b, d, e, xoff, yoff]
+        affine = np.append(t[0:2, 0:2].ravel(), t[0:2, -1])
+        # Use shapely to apply the affine tranformation
+        transformed = affinity.affine_transform(ShapelyPolygon(coords).exterior, affine)
+        coords = np.array(transformed.coords)
+
+    # Transform from 3DTiles SRID to Database SRID
+    # Convert the coordinates using pyproj to database SRID from source SRID
+    # This might work on proj>=8 and pyproj>=3.1 - large_image_wheels gives us these versions
+    src = pyproj.CRS(f'EPSG:{srid}')
+    dest = pyproj.CRS(f'EPSG:{DB_SRID}')
+    project = pyproj.Transformer.from_crs(src, dest, always_xy=True).transform
+    result = shapely_transform(project, ShapelyPolygon(coords))
+    logger.info(f'the coords: {list(result.exterior.coords)}')
+
+    outline = Polygon(list(result.exterior.coords), srid=DB_SRID)
```

Comment:
I'm running into a rather big roadblock here.

The 3D Tiles format provides an affine transform matrix that needs to be applied to the coordinates which are given in a local coordinate system (almost always). Trouble is, I didn't see a clear way to perform this transform with pyproj and resorted to shapely which only works for 2D transforms.

The problem is that 3D tiles explicitly use EPSG 4978/4979 ([reference](https://github.com/CesiumGS/3d-tiles/tree/main/specification#coordinate-reference-system-crs)) which are both earth-centered, earth-fixed (ECEF) reference frames. ECEF reference frames require the elevation datum to properly transform back to EPSG 4326. You can see this issue in [this example I made on epsg.io](http://epsg.io/transform#s_srs=4978&t_srs=4326&x=1213958.8112409&y=-4737648.1118417) where the latitude is always 0:

![Screen Shot 2021-12-07 at 6 57 56 PM](https://user-images.githubusercontent.com/22067021/145134884-6b22dc88-ad53-45f2-ae5c-1b1f159d70ed.png)

There is a post [here](https://gis.stackexchange.com/questions/386070/converting-from-epsg4978-to-epsg4326-using-pyproj-always-produces-0-0-latitude) that mentions this ECEF issue but there is also [this thread](https://issueexplorer.com/issue/geopandas/geopandas/2044) where users report it not being an issue with pyproj 3.1.0 with PROJ 8.1.

In my local configuration, I have pyproj 3.3.0.1 and PROJ 8.2.0 but I'm unable to get this to work and always end up with 0 for the latitude.

@manthey, would you have a chance to look this over?

------

Here is a simpler example to run/test outside of Django/RGD:

```py
from shapely.geometry import Polygon as ShapelyPolygon
import pyproj

from shapely.geometry import Polygon, shape
from shapely.ops import transform

def trans(data, src, dest, always_xy=False):
    src = pyproj.CRS(src)
    dest = pyproj.CRS(dest)
    project = pyproj.Transformer.from_crs(src, dest, always_xy=always_xy).transform
    return transform(project, data)

data = shape(truth)
output = trans(data, 'EPSG:4326', 'EPSG:4978', always_xy=True)
```

```py
>>> list(output.exterior.coords)
[(1213958.8112408777, -4737648.111841724),
 (1213509.9461615593, -4735896.351587741),
 (1216578.6996032828, -4735108.9649644485),
 (1217028.6997835222, -4736860.433972068),
 (1213958.8112408777, -4737648.111841724)]
```

The above is effectively what is handled to us from 3D tiles *after* performing the affine transform.

The trick is getting into a more reasonable SRID for our database:

```py
>>> test = trans(output, 'EPSG:4978', 'EPSG:4326', always_xy=True)
>>> list(test.exterior.coords)
[(-75.62794214702885, 0.0),
 (-75.62794214702885, 0.0),
 (-75.5908126959205, 0.0),
 (-75.5908126959205, 0.0),
 (-75.62794214702885, 0.0)]
```

😕  why is latitude always 0?

Reply:
Some updates:

- ECEF is a 3D coordinate system.  It doesn't make sense unless you have 3 coordinates.  The X-Y plane slices through the globe at the equator, which is why latitude is always 0 if Z=0.
- There exists this for 3-D geospatial transforms: https://gitlab.kitware.com/mattbrown11/colmap_processing/-/blob/master/src/colmap_processing/geo_conversions.py
E.g., rmat_enu_ecef and rmat_ecef_enu could be used to build 4x4 affine matrix (would need to populate the translation in the 4rth column)
- This VTK code does the conversion (from any CRS to ECEF) in C++ with proj. pyproj should have something similar and should work the other direction if you just revert the order of the CRSs https://gitlab.kitware.com/vtk/vtk/-/blob/master/IO/Cesium3DTiles/TreeInformation.cxx#L382